### DISCUSSION TITLE
User feedback

### DISCUSSION THREAD

**Discussion by simoncozens** [MFEK/glif#217] [General]
OK, I'm putting myself in the position of a designer looking at this thing for the first time and trying to draw glyphs with it. This is going to be very harsh because it's about all the things which don't match my expectation, so I'm sorry about that. This is a really cool piece of code, but it needs lots of polishing yet.

<hr>

Right, after installing the Vulcan SDK (oops), I've managed to load up the example. Row of tools down the side: pan, pointer, something like a stethoscope, ... hmm, not sure. These probably could do with tooltips.

The ones I'm most likely to use to edit glyphs are the pointer and the pen. Why is the pen buried in the middle? That should probably be right near the top.

I'm on the Mac and would expect to use standard mouse gestures to move around: pan with two fingers, pinch and spread to zoom in and out, but that doesn't work. Double click to zoom when I'm using the pan tool? No, not that either.

OK, so I can zoom with = and -, that would have been my next guess. But the origin of the zoom is wrong; it's always the top left of the glyph, regardless of my mouse pointer. If I'm focusing on a particular point or area, I want the origin of the zoom to be based on that so I can see it more clearly without having to pan around again.

Let's try pulling some points around. Double-click... no, I'm still in the pan tool. (The toolbar colouring's a bit weird - light grey is an "unselected" colour in my mind and blue is a "primary" but all the tools are blue apart from one light grey one.) If I double click a point, it's because I want to edit it, and so I would expect it to switch to the pointer tool for me. Switch manually, click on a point in the top right corner and... oh, now there's a great big panel over the top of where I just clicked and now I can't see anything any more. Urgh, back to pan tools so I can see the point again, and then back to pointer.

What's this blooming panel actually telling me anyway? I don't really need such a big panel giving me the coordinates of the point and its surrounding offcurves to two decimal places. *Especially* since this is a line segment and there aren't any offcurves, but it seems to be telling me there are some anyway:

<img width="383" alt="Screenshot 2021-12-07 at 13 00 28" src="https://user-images.githubusercontent.com/106728/145033702-d41c2824-308d-4cf4-9669-43ac28ed8f66.png">

I don't need any of that information; maybe a tiny little box where I can adjust the coordinates by hand if I need to, but certainly not something taking up a third of the main window.

Something's not quite right with selection, I seem to have to click about ten pixels below and to the right of the point in order to drag it. Maybe I can select it and nudge it around with keyboard arrows? Oh, no, that doesn't work either.

OK, let's try dragging it... can I shift drag to move it orthogonally from its original point? Ah, no. Ctrl-drag, option-drag? Not that either. Oh, bother, I didn't want to put it over there. Aargh, undo? Command-Z... Nope. Ummm, Edit menu? Wait, there are No menus. OK... Developer uses Linux, so... Ctrl-Z? Phew, yes!

All right, let's try drawing. Hey, that drawing interface is really nice and smooth. Like it. Oops, need to edit an off curve - argh, only one handle moves at a time. I would *really* expect to maintain G1 continuity while dragging the offcurves of a smooth point. Or maybe there's a way to turn that into a smooth point again...? Can't find one. Maybe this is where I need that point panel... hmm, maybe click on "colocated"? Aah, no that turns it into a line segment! Give me my off curve back - urgh, clicking on colocated again doesn't do anything. How do I add an off curve to a path? OK, back to the pen tool maybe? If I hover over the path it seems to give me something like....

Oh.

<img width="589" alt="Screenshot 2021-12-07 at 13 13 55" src="https://user-images.githubusercontent.com/106728/145035601-e495dc96-71c7-445e-bfa8-4f5c9dac3ec6.png">

**Comment by ctrlcctrlv** [comment]
Thanks for trying it out, given all the fixes in #216, including adding nudging points with arrows, I'd encourage you to try again after its merger….which will be tonight, given it just passed CI. I also believe I fixed that crash, but it's hard to say if it's the same one. I fixed a lot of selection/pen crashes.

**Reply by ctrlcctrlv** [reply]
66ca003098e67c5f7697544d24e7ec149f9cd6ed is merged :D

**Comment by ctrlcctrlv** [comment]
Regarding your feedback where you figured out I'm expecting Control key when you remembered I use Linux, this is something I thought about.

Note this file: https://github.com/MFEK/glif/blob/master/resources/default_keymap.xml

The reason I have it is because I was indeed thinking about how Mac users expect a different key than control for general hotkeys. I actually believe it might work right now if you change these lines…

https://github.com/MFEK/glif/blob/5448d1bba9eed00dab155cc1763eef4dd8979eb7/resources/default_keymap.xml#L51-L52

I believe that if you `mod` to `SuperMod` it might work. But, you should know, this is not regularly (ever, really) tested on Mac OS. If CI is building, I test it like once every six months. I'd like that to change, of course, and hope community can help me with that as this software sees more use outside my projects.

I can answer your questions as to how to do it in the current interface, but no, you're just right, the interface needs changed because my answers aren't satisfying. I appreciate you bringing it to my attn again because I'd just gotten used to it, so it's good to see fresh eye experience of user.

**Comment by ctrlcctrlv** [comment]
Oh, and if you confirm that _SuperMod_ is correct, I will add an alias called _AppleMod_ so we can be very sure about it. On startup MFEKglif tells you where the config directory is that it puts your user hotkeys in.

This code does the translation of `<binding mod>`:

![image](https://user-images.githubusercontent.com/838783/145698619-b3ccdac5-b89a-44cf-bff9-9c82f7aa5905.png)

**Comment by ctrlcctrlv** [comment]
FYI I moved this to a discussion because it's not about any one specific bug but just a constellation of issues stemming from MFEKglif's general newness. :)

**Comment by ctrlcctrlv** [comment]
https://github.com/MFEK/glif/commit/939666a9565105a626a2a451775cd849acb414e4

> # Rewrite/simplify MoveHandle tool event & Follow behaviors 
> This deals with the most important elements raised by @simoncozens in
> issue #217. I feel that the rest should be dealt with through
> education/documentation/patching Pen/Select tool related bugs.
